Add indication of Sort order and filter to summary bar.

Bob-IT opened the pull request:
Again not sure if this should go to maint or master but will wait for feedback.
I think this has been asked for several times and there is probably a bug for it that I will try and find. All seems to work, some text may need changing but welcome thoughts.

gjanssens commented on gnucash/gnome/gnc-plugin-page-register.c:
```diff
@@ -2659,6 +2669,114 @@ gnc_plugin_page_register_filter_response_cb (GtkDialog *dialog,
+        gchar *text_r = gpp_get_cleared_match_filter_text (priv->fd.cleared_match, 0x04, _("Reconciled"));
+        gchar *text_f = gpp_get_cleared_match_filter_text (priv->fd.cleared_match, 0x08, _("Frozen"));
+        gchar *text_v = gpp_get_cleared_match_filter_text (priv->fd.cleared_match, 0x10, _("Voided"));
+        text_cleared = g_strconcat (text_u, "\n", text_c, "\n", text_r, "\n", text_f, "\n", text_v, NULL);
```
Let's try to shorten this output. For example make two lines, one with "Show a, b, d" and the second with "Hide c, e, f"

Bob-IT replied:
Will have a look at trying this format, maybe put the dates on one line also ?

gjanssens replied:
Sure, you can try that.

gjanssens commented on gnucash/gnome/gnc-split-reg.c:
```diff
@@ -546,6 +547,63 @@ gsr_redraw_all_cb (GnucashRegister *g_reg, gpointer data)
+        if (gsr->sort_rev)
+        {
+            gtk_style_context_add_class (context, "summary-highlight");
+            gtk_widget_set_tooltip_text (GTK_WIDGET(gsr->sort_label), _("Descending"));
```
It would be even nicer if the Ascending/Descending information was directly visible in the summary bar. I understand the full text would be too long. However perhaps you can add an image of an up-arrow or down-arrow next to the label (right-aligning the label to ensure the text always sticks to the arrow image). We are using arrow images in the register code also. It would be nice to use the same one for consistency.

Bob-IT replied:
Originally I was looking to use a font character 'arrow' to do the same but could not find a suitable one. Now you think it is a reasonable idea I will have look at adding an arrow.

Bob-IT commented on gnucash/gtkbuilder/dialog-preferences.glade:
```diff
@@ -58,6 +58,12 @@
   </object>
+  <object class="GtkAdjustment" id="gl_number_of_days_adj">
+    <property name="lower">30</property>
+    <property name="upper">1100</property>
```
I had thought about this and I was wondering if these settings should be saved to the .gcm file along with the other register settings which is where the GUID 000...000 is being saved. I assume this would require a feature option for the change from account file to .gcm file.
With an unfiltered GL, wouldn't that mean all transactions would be loaded!

The good thing about setting a number is that it is a rolling period so you always see 'x' amount of days. There could be another filter option to enable this for all registers with each one specifying there own limit then there would be no need for the preference option.

Now I am not sure how much to do in this PR?

gjanssens replied:
Ah, I see what you mean. The current hard-coded 30 days is converted in a dynamic starting date filter for the GL.

Without a filter would indeed mean all transactions would be loaded. I believe it's uncommon to require this, but not unthinkable. That's why it should not be the default for the General Journal, and the current 30 days sounds like a good default.

I'm not sure I understand what you mean with the change from account file to .gcm file. Do we store filter information in the account file ?

 I think allowing the user to specify a custom range would be a good thing for all registers. And I would prefer that over setting a global range for the GL only.

Bob-IT replied:
The filter and sort order are stored as KVP string entries for the account like...
 <slot:key>filter</slot:key>  <slot:value type="string">0x001d,0,0</slot:value>
 <slot:key>sort-order</slot:key>  <slot:value type="string">BY_DATE_ENTERED</slot:value>

gjanssens replied:
Right. It's better to be consistent.
The minimal change would then be to store the sort order and filter in kvp as well. As the GL is not a separate account we have to find a different spot to attach this. Two possible objects can be candidates: the book account itself (there's only one GL for a book) or the root account as that one is hidden so it doesn't need filter and sort keys for itself. And in fact I like the root account best as it most closely matches the GL. If we were to display the root account with sub accounts in multi-split view it would resemble the GL pretty well.
In both cases this would modify the data format, so this particular change should go on master. And in this case I would even skip the Feature code and backwards compatibility stuff; the GL on maint won't use the kvp, but retain it if it exists. So switching between maint and master in the future will have the new option in effect in master while maint continues to behave as it does now and yet it won't interfere with master's new kvp.

Related to this I still have my reservations about storing presentation stuff in the data file directly. Or it requires a serious redesign to separate it from the data properly inside the db/xml file. It really complicates the data structure. If such a restructuring is not a feasible medium term goal we should continue to move the presentation settings to the book's metadata file (.gcm). In this context that would mean moving all the kvp data you refer to into the gcm file. And in that case we will need migration and compatibility guards in place.

Hmm, perhaps stick with the minimal effort option for now until we're ready to revisit the data/presentation separation properly...

Bob-IT commented:
OK, I have made some changes and all looks good to me but lets see. Added an arrow and filter tooltip changed along with showing the Filter label. I have also added saving the settings to the .gcm file as well as the account kvp and loading them from the .gcm file falling back to the account kvp. Added a feature for the full migration to just using .gcm file. If I understand correctly if this gets accepted on maint and master rebased on that I can then do the migration code on master along with some tidy up.

gjanssens commented on gnucash/gnome/gnc-plugin-page-register.c:
```diff
@@ -2659,6 +2669,148 @@ gnc_plugin_page_register_filter_response_cb (GtkDialog *dialog,
+
+        gboolean first_show = TRUE, first_hide = TRUE;
+
+        if ((priv->fd.cleared_match & 0x01) == 0x01)
```
Also gpp_**update**_match_filter_text would probably be a better name.

Bob-IT commented on libgnucash/engine/gnc-features.c:
```diff
@@ -48,6 +48,7 @@ static gncFeature known_features[] =
     { GNC_FEATURE_GUID_BAYESIAN, "Use account GUID as key for Bayesian data (requires at least GnuCash 2.6.12)" },
     { GNC_FEATURE_GUID_FLAT_BAYESIAN, "Use account GUID as key for bayesian data and store KVP flat (requires at least Gnucash 2.6.19)" },
     { GNC_FEATURE_SQLITE3_ISO_DATES, "Use ISO formatted date-time strings in SQLite3 databases (requires at least GnuCash 2.6.20)"},
+    { GNC_FEATURE_REG_SORT_FILTER, "Store the Register sort and filter settings in .gcm file only (requires at least GnuCash 4.0)"},
```
Not sure about this, with these changes the sort and filter are saved to both .gcm and kvp entries with loading from .gcm and then kvp if no entry found. So in master leading to 4.0 I would set this feature to true when a conversion routine ran moving entries from kvp to .gcm and a change to only saving and loading from .gcm file.

Bob-IT replied:
So all I need to do for this PR is amend the feature text and push that along with the other changes. Then wait for 3.3 and do the conversion, delete old keys and set the feature flag.

gjanssens replied:
No. Gnucash 3.x should **never** set the feature flag by itself because that would mean older early 3.x version and 2.6.21 won't be able to work on such a book any more. We try to maintain this backwards compatibility to the last release of the previous stable series.

But gnucash 3.3 and up should behave differently if it's set.

It's a difficult matter to do right. And while thinking this through again I believe the behaviour should still be slightly different than what I suggested.
Perhaps the easiest way to visualize it is that the feature flag's presence defines the storage format of the filter/sort stuff. If the flag is absent, filter/sort should go in kvp. If the flag is set, filter/sort should go in .gcm. No version of gnucash should ever write to both formats at the same time.

So what format 3.x will use depends on the feature flag being present or not. If the flag is set, 3.3 and up should only interact with .gcm. If the flag is not set, 3.x should only interact with kvp (which works for the complete 3.x series and even 2.6.21).

So 3.x should be able to read and write both kvp and .gcm, but never at the same time. The feature flag defines which format to use.

The conversion from kvp to .gcm can only happen in a 4.x release, the first time a book without feature flag set is opened. Again no 3.x release should ever do the conversion.

That should allow you to cleanly separate the two implementations. In 3.3 you add both implementations and select one or the other depending on the feature flag (you may have to extract saving and loading of filters to separate functions to make this easier, I didn't recheck).

In 4.0 you only use the kvp functions in a one-time migration function. This will
1. to read the old values from kvp
1. clear the kvp
1. write the values to .gcm
1. set the feature flag

Bob-IT replied:
Will sleep on this....

Bob-IT commented:
I have made changes locally and will push once the feature query is iron out.

Bob-IT commented:
Made some changes which address the filter text creation and the feature setting.
With these changes the sort/filter settings will be saved to both .gcm and kvp as the GL settings are only saved/restored from the .gcm file. With the feature set, only the .gcm file will be used. So with version 4.0/master the feature will be set, conversion of all kvp entries to .gcm which by that time will probably be a check for .gcm entry and delete the kvp entries. Should the file get opened by a version 3.3 and above only the .gcm will be used. For less than 3.3, the feature warning should be displayed.

gjanssens commented:
Thanks for the update. However I still don't think this is conceptually right. As implemented now gnucash 3.3 and up will be reading and writing both from gcm and kvp (until 4.x sets the feature flag the first time). This opens the door for all kinds of out of sync issues. Let's assume a user has GC 3.3 on one pc and 3.2 on another.
* The user loads a book in 3.3 and sets a filter/sort => this will be written both to kvp and gcm.
* Then switches to 3.2 => 3.2 will load from kvp and ignore gcm
* In 3.2 the user decides on another filter/sort => the new filter will only be saved to kvp
* User switches back to 3.3 => 3.3 will load old filter/sort from gcm because this takes precedence
=> user annoyance

This can be avoided by making 3.3 only use kvp unless the flag gets set at which point it should only use gcm.

I do understand this makes it a bit more complicated for GL as you have only implemented this in gcm. To deal with this you can either make GL filter loading/saving an exception (in that that one always uses gcm only) or you can implement GL filter storage in kvp for the 3.x series and just let it go through the same conversion as all the other in the first 4.x run.
Or user access to GL filter storage gets delayed until 4.x completely.
These are IMO the only ways to guarantee the user never ends up with and outdated filter/sort. Which one to choose depends a bit on the complexity of the code.

I suspect the first option is the easiest to implement. You may need to rearrange the filter load and save code a bit to have a construct like this pseudo code:
```
if (ledger_type == GL)
   load from gcm
else if (gcm_feature_flag_set)
  load from gcm
else
  load from kvp
```
And the same for saving of course.

gjanssens commented:
Thanks for the update. I'm fine now with the concept. However when testing your code, there is still a bug with saving the filter for the General Journal.
Setting the filter to either "Show all" or "Select Range: Earliest/Latest" this filter will never be stored. This is the correct case for normal registers. However for the General Journal the default filter in case no filter is set will be "Show number of days: 30". So for the General Journal this default filter test upon saving should be amended.

Bob-IT commented:
Slightly confused about your last statement, the default_filter value is never stored but used to erase the existing value if present. On load, if no filter present it will get the default_filter value but for the LD_GL it is checked and adjusted to 30 days. This can be seen starting at line 1222 of gnc_plugin_page_register.

gjanssens commented:
Try these exact steps, perhaps that will make it clearer what I mean:

1. start gnucash
2. open the General Journal
=> Unless you had already saved a filter, it will show only the last 30 days.
3. select View->Filter by... and set filter to "Show all"
4. tick box to save filter and hit ok
=> The GJ is now showing all transactions
5. Save and restart gnucash
=> The GJ opens with showing only the last 30 days.
Expected  behaviour: the GJ opens with all transactions as that's what the user saved.

I suspect this is due to what you consider the default when deciding what to write to the .gcm file. It looks like you unconditionally assume there the default is "Show all". However for the GJ the default is "Show last 30 days". So if the register at hand is the GJ, nothing should be written if the current filter is "Show last 30 days" where for all other registers nothing should be written if the current filter is "Show all".

It may also be the load step is wrong though, I haven't checked in detail. Maybe you do save a filter to file for the GJ when "Show all" is chosen (though I didn't find one). In that case the load step should detect a filter is set and not convert it to "Show last 30 days" in case of the GJ.

Bob-IT commented:
OK, I think I have sorted every thing, tested with various combinations and all looked how I expected it to look. Also tried setting feature and all wrote to the.gcm file as expected.

gjanssens commented on gnucash/gnome/gnc-plugin-page-register.c:
```diff
@@ -1860,13 +1870,21 @@ gnc_plugin_page_register_set_filter (GncPluginPage *plugin_page, const gchar *fi
+        default_num_str = "0";
+
+    default_filter = g_strdup_printf("%s,%s,%s,%s", DEFAULT_FILTER, "0", "0", default_num_str);
+
```
Thanks for the update. It works as advertised now.

Your most recent change on the other hand now introduces a maintenance issue: it uses the same two "magic" number in 3 different locations (30 and 0).

Please factor this test out in a separate internal function that takes the ledger type as input and returns the filter_default_days as a const char*.

That will allow you to create default_filter directly (as it was) like
```const gchar *default_num_str = ledger_default_filter_days (ledger_type);
gchar *default_filter = g_strdup_printf("%s,%s,%s,%s", DEFAULT_FILTER, "0", "0", default_num_str);```

You can use #define's or const char *'s at the top of the source to setup sensible names for the two magic numbers themselves.

Bob-IT replied:
Have made the changes, did not see that at all.

Bob-IT commented:
Made changes as above and added another commit which I think is appropriate to reduce the number of time the ledger display is refreshed. I could add it to a separate PR but as in the same area I thought it would be easier here.

gjanssens commented:
Thanks for the update. The final commit is ok as is and hence I have merged your branch. There was a hiccup during code->github synchronisation so the changes will only appear on github after the first future commit.

gjanssens commented:
Oh, yeah, and thanks for your persistence :)

Bob-IT commented:
Yippee, one day I may get better at this, here's hoping...

gjanssens commented:
Oh well, it's always easier to review than to write, so you did most of the work here!

gjanssens commented:
By chance I saw today that meanwhile you're in the [top 10 of all time gnucash contributors](https://www.openhub.net/p/gnucash/contributors?page=1&query=&sort=commits&time_span=) at openhub. Congratulations !

jralls commented:
And you're ahead of me by ~370 commits... but we've both got a way to go to catch Linas! ;-)
